netwaystev2/src/filter/interface.rs

```diff
@@ -0,0 +1,242 @@
+        // TODO: limit chats and game_updates based on MTU!
+        chats:           Vec<BroadcastChatMessage>, // All non-acknowledged chats are sent each update
+        game_update_seq: Option<u64>,
+        /* PR_GATE add in later
+        game_updates:    Vec<GameUpdate>, // Information pertaining to a game tick update.
+        universe_update: UniUpdate,       // TODO: add support
+        */
+        ping:            PingPong, // Used for server-to-client latency measurement (no room needed)
+    },
+    UpdateReply {
+        // in-game: sent by client in reply to server
+        cookie:               String,
+        last_chat_seq:        Option<u64>, // sequence number of latest chat msg. received from server
+        last_game_update_seq: Option<u64>, // seq. number of latest game update from server
+        last_full_gen:        Option<u64>, // generation number client is currently at
+        /* PR_GATE add in later
+        partial_gen:          Option<GenPartInfo>, // partial gen info, if some but not all GenStateDiffParts recv'd
+        */
+        pong:                 PingPong, // Used for server-to-client latency measurement
+    },
+    GetStatus {
+        ping: PingPong, // Used for client-to-server latency measurement
+    },
+    Status {
+        pong:           PingPong, // used for client-to-server latency measurement
+        server_version: String,
+        player_count:   u64,
+        room_count:     u64,
+        server_name:    String,
+        // TODO: max players?
+    }, // Provide basic server information to the requester
+}
+
+impl Default for Packet {
+    fn default() -> Self {
+        Packet::Request {
+            sequence:     0,
+            response_ack: None,
+            cookie:       None,
+            action:       RequestAction::None,
+        }
+    }
+}
+
+// chat messages sent from server to all clients other than originating client
+#[derive(Serialize, Deserialize, Debug, Clone)]
+pub struct BroadcastChatMessage {
+    pub chat_seq:    Option<u64>, // Some(<number>) when sent to clients (starts at 0 for first
+    // chat message sent to this client in this room); None when
+    // internal to server
+    pub player_name: String,
+    pub message:     String, // should not contain newlines
+}
+
+impl BroadcastChatMessage {
+    #[allow(unused)]
+    pub fn new(sequence: u64, name: String, msg: String) -> BroadcastChatMessage {
+        BroadcastChatMessage {
+            chat_seq:    Some(sequence),
+            player_name: name,
+            message:     msg,
+        }
+    }
+
+    fn sequence_number(&self) -> u64 {
+        if let Some(v) = self.chat_seq {
+            v
+        } else {
+            0
+        }
+    }
+}
+
+#[derive(Serialize, Deserialize, PartialEq, Debug, Clone)]
+pub struct RoomList {
+    pub room_name:    String,
+    pub player_count: u8,
+    // TODO: add support
+    pub in_progress:  bool,
+}
+
+pub enum FilterEndpointData {
```

Comment:
What do you think about putting this somewhere else? I feel like this and the error below it don't belong in `interface.rs` since they are internal to the Filter.

Reply:
Yep agreed! There's a bunch of stuff here that doesn't belong -- rather, a lot of it is tailed towards the app-layer. I'll move it out.

Reply:
I created a `src/protocol` folder which contains the data structure definitions for `Packet` and the like. Let me know what you think!

Reply:
Wasn't sure if I should move `PingPong` too but I decided to leave it in `filter.rs` for now.

Reply:
I really like the organization in `src/protocol`. I can actually find structs and enums with this.

The only thing that I would change is this:

```rust
        /* PR_GATE add in later
        game_updates:    Vec<GameUpdate>, // Information pertaining to a game tick update.
        universe_update: UniUpdate,       // TODO: add support
        */
```
Can you make the complete structure we have in the `netwayste` crate be here too (no temporarily commented data structures) please?

Also I am not sure why we need a `Default` implementation for `Packet...?

Reply:
The other thought I had is maybe we could merge `response.rs` into `request.rs`. Otherwise there will need to be an `updatereply.rs` file added to be consistent, and that seems like a lot of files. 😁

Reply:
Removed the `PR_GATE` comments and added in their associated data structures. 

I like that `request.rs` and `response.rs` are separate. The idea was that any data structures used within a `Packet` variant's would be defined in their respective files for quick look up. `UpdateReply` uses only primitive data types so there wasn't a need to define an `updatereply.rs`.... yet.